File: BOJ/Silver III/BOJ2805.py
#BOJ2805 나무 자르기 20210217
import sys
input = sys.stdin.readline

def main():
    n,m = map(int,input().rstrip().split(" "))
    li = sorted(map(int, input().rstrip().split(" ")))
    f = 0; l = li[-1]
    mid = (f+l)//2
    res = 0
    while f <= l:
        mid = (f+l)//2
        s = sum([v-mid if v > mid else 0 for v in li])
        if s >= m:
            f = mid + 1
            res = max(res,mid)
        else:
            l = mid - 1
    print(res)
if __name__ == '__main__':
    main()

File: BOJ/Silver III/BOJ2805.java
//BOJ2805 나무 자르기 20210706
import java.io.*;
import java.util.*;

import static java.lang.Math.*;

public class Main {
    private static final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static final BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    public static void main(String[] args) throws IOException {
        StringTokenizer stk = new StringTokenizer(br.readLine());
        long n = Integer.parseInt(stk.nextToken());
        long m = Integer.parseInt(stk.nextToken());
        long[] arr = Arrays.stream(br.readLine().split(" ")).mapToLong(Long::parseLong).sorted().toArray();

        long ans = 0;
        long l = 0, r = arr[(int) (n-1)];
        while(l <= r){
            long mid = (l+r) / 2;
            long tmp = 0;
            for(long v: arr){
                if(v > mid) tmp += v - mid;
            }
            if(tmp >= m){
                ans = mid;
                l = mid + 1;
            }
            else{
                r = mid - 1;
            }
        }
        System.out.println(ans);
    }
}